util/rowDecoder/decoder.go

```diff
@@ -79,8 +87,29 @@ func (rd *RowDecoder) DecodeAndEvalRowWithMap(ctx sessionctx.Context, b []byte,
 		return row, nil
 	}
 
-	for id, v := range row {
-		rd.mutRow.SetValue(rd.columns[id].Info.Offset, v.GetValue())
+	for _, dCol := range rd.columns {
+		id := dCol.Info.ID
+		val, ok := row[id]
+		c, ok1 := rd.colsInGeneratedExpr[id]
+		if ok || !ok1 {
+			rd.mutRow.SetValue(rd.columns[id].Info.Offset, val.GetValue())
+			continue
+		}
+
+		// Get the default value of the column in the generated column expression.
+		if c.IsPKHandleColumn(rd.tbl.Meta()) {
```

Comment:
How about use `if rd.pkIsHandle && mysql.HasPriKeyFlag(dCol.Info.Flag)`, then we can only store `pkIsHandle` in RowDecoder, and remove `tbl`, `colsInGeneratedExpr`.

Reply:
It seems that we can remove `tbl`, but why we can remove `colsInGeneratedExpr`？
I think storing `tbl` or `pkIsHandle` are both OK, and using`IsPKHandleColumn` is easier.

Reply:
Removed `colsInGeneratedExpr`.